### ISSUE TITLE
REVERSE() on special character results in "Assertion `strcmp(dataptr, normalized) == 0' failed."

### ISSUE THREAD

**Issue by mrigger** [cwida/duckdb#567]
Consider the following statement:

```sql
SELECT REVERSE('S̈a︍'); --  Assertion `strcmp(dataptr, normalized) == 0' failed.
```
Unexpectedly, the `SELECT` results in an assertion error:
```
/duckdb/src/common/types/string_type.cpp:18: void duckdb::string_t::Verify(): Assertion `strcmp(dataptr, normalized) == 0' failed.
```
I found this based on commit cc5e3f29bb1b4ea1acf185b19b57b5b009b33067.

**Comment by Mytherin** [comment]
Thanks for the report! An excellent bug report, it has triggered some discussion.

The problem is a bit complex to deal with. The bug is in the REVERSE function not properly handling combining unicode characters. In regular situations we normalize those (to NFC form), but for the character S̈ there is no such normalized form. The reverse of `S̈a` then creates `äS`, but with the combining character still in there. However, since there is a single code point variant of ä, the NFC normalization process modifies that character and triggers this assertion.

To correctly handle combining characters in the REVERSE function basically requires us to pull in the ICU library, as handling this ourselves is very complex and error prone. To fix the assertion itself, we could also just run the normalize again after the reverse, but this would incorrectly lead to the result `äS` being produced. I have checked, though, and this is consistent with other RDBMS systems (Postgres/MySQL) and Python: they all also just reverse code points. 

The question is also perhaps do we need to reverse arbitrary unicode strings correctly in the first place, and what is the use case that we are handling there? Is it fine to just reverse code points?

**Comment by Mytherin** [comment]
Actually we have decided to fix this properly after all in 2927f7835bdee40695b8cf946890cca86c42623a. The UTF8proc library that we are using has support for grapheme clusters, so it was not so tricky to implement as we thought :)

Now `SELECT REVERSE('S̈a︍');` correctly produces `aS̈` as expected. Works properly with tons of unicode combinatorial things, e.g.:

```sql
 SELECT REVERSE('Z͑ͫ̓ͪ̂ͫ̽͏̴̙̤̞͉͚̯̞̠͍A̴̵̜̰͔ͫ͗͢');
--A̴̵̜̰͔ͫ͗͢Z͑ͫ̓ͪ̂ͫ̽͏̴̙̤̞͉͚̯̞̠͍
```
